File: SpectraLoggerUI/Sources/SpectraLoggerUI/ViewModels/LogsViewModel.swift
import SwiftUI
import SpectraLogger

/// ViewModel for the Logs screen, bridging SwiftUI to KMP storage
@Observable
@MainActor
final class LogsViewModel {
    var logs: [LogEntry] = [] {
        didSet { updateFilteredLogs() }
    }

    var filteredLogs: [LogEntry] = []
    var isLoading = true

    var searchText = "" {
        didSet { updateFilteredLogs() }
    }

    var selectedLevels: Set<LogLevel> = [] {
        didSet { updateFilteredLogs() }
    }

    /// Advanced filters from Filter Screen
    var advancedFilter = LogsFilter() {
        didSet { updateFilteredLogs() }
    }

    var availableTags: [String] = []
    
    /// Check if any filters are active (levels or advanced)
    var hasActiveFilters: Bool {
        !selectedLevels.isEmpty || advancedFilter.hasActiveFilters
    }
    
    /// Total count of all active filters for badge display
    var totalActiveFilterCount: Int {
        selectedLevels.count + advancedFilter.activeFilterCount
    }

    private let storage: LogStorage
    private var filterTask: Task<Void, Never>?

    init(storage: LogStorage = SpectraLogger.shared.logStorage) {
        self.storage = storage
        loadLogs()
    }

    private func updateFilteredLogs() {
        // Cancel previous task to debounce rapid changes
        filterTask?.cancel()

        // Schedule new filtering task with debounce
        filterTask = Task {
            try? await Task.sleep(nanoseconds: 300_000_000) // 300ms debounce

            if !Task.isCancelled {
                applyFilters(
                    searchText: searchText,
                    levels: selectedLevels,
                    advancedFilter: advancedFilter,
                    logs: logs
                )
            }
        }
    }

    func loadLogs() {
        Task {
            isLoading = true
            do {
                // Query logs from KMP storage with no filter (all properties nil) and no limit
                let noFilter = LogFilter(levels: nil, tags: nil, searchText: nil, fromTimestamp: nil, toTimestamp: nil)
                let result = try await storage.query(filter: noFilter, limit: nil)
                DispatchQueue.main.async {
                    self.logs = result
                    // Extract unique tags from logs, sorted alphabetically
                    self.availableTags = Array(Set(result.map { $0.tag })).sorted()
                    self.applyFilters(
                        searchText: self.searchText,
                        levels: self.selectedLevels,
                        advancedFilter: self.advancedFilter,
                        logs: result
                    )
                    self.isLoading = false
                }
            } catch {
                print("Error loading logs: \(error)")
                DispatchQueue.main.async {
                    self.isLoading = false
                }
            }
        }
    }

    private func observeNewLogs() {
        // TODO: Implement log observation when KMP Flow-to-Swift interop is available
        // For now, logs are loaded on demand with loadLogs()
    }

    private func applyFilters(
        searchText: String,
        levels: Set<LogLevel>,
        advancedFilter: LogsFilter,
        logs: [LogEntry]
    ) {
        var filtered = logs

        // Filter by log level
        if !levels.isEmpty {
            filtered = filtered.filter { levels.contains($0.level) }
        }

        // Filter by tags (from advanced filter)
        let allTags = advancedFilter.allSelectedTags
        if !allTags.isEmpty {
            filtered = filtered.filter { allTags.contains($0.tag) }
        }

        // Filter by time range
        if let fromTimestamp = advancedFilter.fromTimestamp {
            let fromEpoch = fromTimestamp.timeIntervalSince1970
            filtered = filtered.filter { log in
                let logEpoch = Double(log.timestamp.epochSeconds)
                return logEpoch >= fromEpoch
            }
        }
        
        if let toTimestamp = advancedFilter.toTimestamp {
            let toEpoch = toTimestamp.timeIntervalSince1970
            filtered = filtered.filter { log in
                let logEpoch = Double(log.timestamp.epochSeconds)
                return logEpoch <= toEpoch
            }
        }

        // Filter by metadata
        if !advancedFilter.metadataKey.isEmpty && !advancedFilter.metadataValue.isEmpty {
            filtered = filtered.filter { log in
                if let value = log.metadata[advancedFilter.metadataKey] {
                    return value.localizedCaseInsensitiveContains(advancedFilter.metadataValue)
                }
                return false
            }
        }

        // Filter by has error
        if advancedFilter.hasErrorOnly {
            filtered = filtered.filter { log in
                log.throwable != nil || log.metadata["stack_trace"] != nil
            }
        }

        // Filter by search text (min 2 chars)
        if searchText.count >= 2 {
            filtered = filtered.filter { log in
                log.message.localizedCaseInsensitiveContains(searchText) ||
                log.tag.localizedCaseInsensitiveContains(searchText) ||
                log.level.name.localizedCaseInsensitiveContains(searchText)
            }
        }

        filteredLogs = filtered
    }

    func toggleLevel(_ level: LogLevel) {
        if selectedLevels.contains(level) {
            selectedLevels.remove(level)
        } else {
            selectedLevels.insert(level)
        }
    }

    /// Remove a specific tag from the filter
    func removeTagFilter(_ tag: String) {
        advancedFilter.selectedTags.remove(tag)
        advancedFilter.customTags.remove(tag)
    }

    /// Clear time range filter
    func clearTimeRangeFilter() {
        advancedFilter.fromTimestamp = nil
        advancedFilter.toTimestamp = nil
    }

    /// Clear metadata filter
    func clearMetadataFilter() {
        advancedFilter.metadataKey = ""
        advancedFilter.metadataValue = ""
    }

    /// Clear has error filter
    func clearHasErrorFilter() {
        advancedFilter.hasErrorOnly = false
    }

    func clearLogs() {
        Task {
            try await storage.clear()
            logs = []
            filteredLogs = []
            availableTags = []
        }
    }
}

File: spectra-ui-android/src/main/kotlin/com/spectra/logger/ui/compose/LogsViewModel.kt
package com.spectra.logger.ui.compose

import android.content.Context
import android.content.Intent
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.spectra.logger.SpectraLogger
import com.spectra.logger.domain.model.LogEntry
import com.spectra.logger.domain.model.LogLevel
import com.spectra.logger.domain.storage.LogFilter
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.datetime.Instant

/**
 * ViewModel for the Logs screen
 */
class LogsViewModel : ViewModel() {
    
    private val storage = SpectraLogger.logStorage
    
    private val _uiState = MutableStateFlow(LogsUiState())
    val uiState: StateFlow<LogsUiState> = _uiState.asStateFlow()

    init {
        loadLogs()
    }

    fun loadLogs() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            try {
                val noFilter = LogFilter(
                    levels = null,
                    tags = null,
                    searchText = null,
                    fromTimestamp = null,
                    toTimestamp = null
                )
                val logs = storage.query(filter = noFilter, limit = null)
                val availableTags = logs.map { it.tag }.distinct().sorted()
                
                _uiState.update { state ->
                    state.copy(
                        logs = logs,
                        availableTags = availableTags,
                        isLoading = false
                    )
                }
                applyFilters()
            } catch (e: Exception) {
                _uiState.update { it.copy(isLoading = false) }
            }
        }
    }

    fun onSearchTextChanged(text: String) {
        _uiState.update { it.copy(searchText = text) }
        applyFilters()
    }

    fun toggleLevel(level: LogLevel) {
        _uiState.update { state ->
            val newLevels = if (state.selectedLevels.contains(level)) {
                state.selectedLevels - level
            } else {
                state.selectedLevels + level
            }
            state.copy(selectedLevels = newLevels)
        }
        applyFilters()
    }
    
    fun updateLevels(levels: Set<LogLevel>) {
        _uiState.update { it.copy(selectedLevels = levels) }
        applyFilters()
    }

    fun updateFilter(filter: AdvancedFilter) {
        _uiState.update { it.copy(advancedFilter = filter) }
        applyFilters()
    }

    fun removeTagFilter(tag: String) {
        _uiState.update { state ->
            state.copy(
                advancedFilter = state.advancedFilter.copy(
                    selectedTags = state.advancedFilter.selectedTags - tag,
                    customTags = state.advancedFilter.customTags - tag
                )
            )
        }
        applyFilters()
    }

    fun clearTimeRangeFilter() {
        _uiState.update { state ->
            state.copy(
                advancedFilter = state.advancedFilter.copy(
                    fromTimestamp = null,
                    toTimestamp = null
                )
            )
        }
        applyFilters()
    }

    fun clearHasErrorFilter() {
        _uiState.update { state ->
            state.copy(
                advancedFilter = state.advancedFilter.copy(hasErrorOnly = false)
            )
        }
        applyFilters()
    }

    fun clearLogs() {
        viewModelScope.launch {
            storage.clear()
            _uiState.update {
                it.copy(logs = emptyList(), filteredLogs = emptyList(), availableTags = emptyList())
            }
        }
    }
    
    fun shareLogs(logs: List<LogEntry>) {
        // TODO: Implement actual sharing via Intent
        // This would require context access, typically done via Application class or passed in
        val logsText = logs.joinToString("\n") { log ->
            "[${log.level.name}] ${log.timestamp} - ${log.tag}: ${log.message}"
        }
        // For now, we just prepare the text. Actual sharing would be done in the UI layer.
        println("Share logs: $logsText")
    }

    private fun applyFilters() {
        _uiState.update { state ->
            var filtered = state.logs

            // Filter by log level
            if (state.selectedLevels.isNotEmpty()) {
                filtered = filtered.filter { state.selectedLevels.contains(it.level) }
            }

            // Filter by tags
            val allTags = state.advancedFilter.allSelectedTags
            if (allTags.isNotEmpty()) {
                filtered = filtered.filter { allTags.contains(it.tag) }
            }

            // Filter by time range
            state.advancedFilter.fromTimestamp?.let { from ->
                filtered = filtered.filter { it.timestamp >= from }
            }
            state.advancedFilter.toTimestamp?.let { to ->
                filtered = filtered.filter { it.timestamp <= to }
            }

            // Filter by has error
            if (state.advancedFilter.hasErrorOnly) {
                filtered = filtered.filter { 
                    it.throwable != null || it.metadata.containsKey("stack_trace")
                }
            }

            // Filter by search text (min 2 chars)
            if (state.searchText.length >= 2) {
                val query = state.searchText.lowercase()
                filtered = filtered.filter { log ->
                    log.message.lowercase().contains(query) ||
                    log.tag.lowercase().contains(query) ||
                    log.level.name.lowercase().contains(query)
                }
            }

            state.copy(filteredLogs = filtered)
        }
    }
}

/**
 * UI State for Logs screen
 */
data class LogsUiState(
    val logs: List<LogEntry> = emptyList(),
    val filteredLogs: List<LogEntry> = emptyList(),
    val isLoading: Boolean = true,
    val searchText: String = "",
    val selectedLevels: Set<LogLevel> = emptySet(),
    val availableTags: List<String> = emptyList(),
    val advancedFilter: AdvancedFilter = AdvancedFilter()
) {
    val hasActiveFilters: Boolean
        get() = advancedFilter.hasActiveFilters
    
    val hasAnyActiveFilters: Boolean
        get() = selectedLevels.isNotEmpty() || advancedFilter.hasActiveFilters
    
    val activeFilterCount: Int
        get() = advancedFilter.activeFilterCount
    
    val totalActiveFilterCount: Int
        get() = selectedLevels.size + advancedFilter.activeFilterCount
    
    val selectedTags: Set<String>
        get() = advancedFilter.allSelectedTags
    
    val hasTimeRangeFilter: Boolean
        get() = advancedFilter.fromTimestamp != null || advancedFilter.toTimestamp != null
    
    val hasErrorOnly: Boolean
        get() = advancedFilter.hasErrorOnly
}

/**
 * Advanced filter configuration
 */
data class AdvancedFilter(
    val selectedTags: Set<String> = emptySet(),
    val customTags: Set<String> = emptySet(),
    val fromTimestamp: Instant? = null,
    val toTimestamp: Instant? = null,
    val metadataKey: String = "",
    val metadataValue: String = "",
    val hasErrorOnly: Boolean = false
) {
    val allSelectedTags: Set<String>
        get() = selectedTags + customTags
    
    val hasActiveFilters: Boolean
        get() = selectedTags.isNotEmpty() ||
                customTags.isNotEmpty() ||
                fromTimestamp != null ||
                toTimestamp != null ||
                (metadataKey.isNotEmpty() && metadataValue.isNotEmpty()) ||
                hasErrorOnly
    
    val activeFilterCount: Int
        get() {
            var count = 0
            if (allSelectedTags.isNotEmpty()) count++
            if (fromTimestamp != null || toTimestamp != null) count++
            if (metadataKey.isNotEmpty() && metadataValue.isNotEmpty()) count++
            if (hasErrorOnly) count++
            return count
        }
}
